feat: Implemented search results page functionality

afanjul opened the pull request:
# Pull Request Template

## Description
Implemented search results page functionality. Now you can press "Enter" to search by term and display results in a results page. Also now you can link to /hc/{account}/en/search?query=XXXXXX to view search results for XXXXXX query.

fixes: https://github.com/chatwoot/chatwoot/issues/10945

## Screenshots

Classic layout search results:
<img width="3840" height="2160" alt="classic-results" src="https://github.com/user-attachments/assets/3bbb3272-33ca-4eb4-b80a-76ed77442088" />

Classic layout pagination:
<img width="3840" height="2160" alt="classic-page-two" src="https://github.com/user-attachments/assets/062b09d3-7c58-4d3b-8611-b94375e7db51" />

Classic layout empty search:
<img width="3840" height="2160" alt="no-results" src="https://github.com/user-attachments/assets/c5e3f47a-cd9a-4e14-ae92-ccba00c89e98" />

Documentation layout search results:
<img width="3840" height="2160" alt="documentation-results" src="https://github.com/user-attachments/assets/9e45d8d9-c975-4589-b6c6-3bc7bb3c588e" />

Documentation layout dark theme:
<img width="3840" height="2160" alt="documentation-dark" src="https://github.com/user-attachments/assets/cdb6ed63-4241-4b32-9f79-7d92ed479fc8" />

Plain embedded dark layout:
<img width="3840" height="2160" alt="plain-embedded-dark" src="https://github.com/user-attachments/assets/7deb02b9-9f24-48fb-8979-a2ecd7002c05" />

afanjul committed:
feat: Implement search results functionality with improved UX in PublicArticleSearch component
- Now you can use /hc/{account}/en/search?query=XXX to view search results

afanjul committed:
Merge branch 'develop' into feature/10945-search-results-page

scmmishra committed:
Merge branch 'develop' into feature/10945-search-results-page

scmmishra committed:
feat: use turbo instead of turbo links

scmmishra committed:
feat: better search UX on the page

scmmishra committed:
feat: get rid of turbo links

scmmishra committed:
Merge branch 'develop' into feature/10945-search-results-page

scmmishra committed:
Merge branch 'develop' into feature/10945-search-results-page

sojan-official committed:
Merge branch 'develop' into feature/10945-search-results-page

scmmishra commented:
Hey @pranavrajs can you take a look at the changes I made

scmmishra committed:
Merge branch 'develop' into feature/10945-search-results-page

scmmishra committed:
Merge branch 'develop' into feature/10945-search-results-page

scmmishra committed:
Merge branch 'develop' into feature/10945-search-results-page

pranavrajs commented:
@scmmishra It looks like the search on the main page and the dedicated search page are not using the same logic. We have both embeddings-based search and regular search, and it seems like the two are getting mixed up in the current PR. Please take a look at the search results below for reference.

<img width="1021" alt="Screenshot 2025-06-02 at 6 42 46 PM" src="https://github.com/user-attachments/assets/dfa1468a-8653-49c7-97c3-cbef3f6b1925" />

<img width="1031" alt="Screenshot 2025-06-02 at 6 42 53 PM" src="https://github.com/user-attachments/assets/7c66d078-aa02-4b3e-a037-539e26b0c8fb" />

pranavrajs commented:
Also, please check this. 

<img width="914" alt="Screenshot 2025-06-02 at 6 45 19 PM" src="https://github.com/user-attachments/assets/4fae5dfc-dae7-4a1b-857c-c64e951453b6" />

scmmishra committed:
feat: add enterprise controller for search

scmmishra committed:
fix: remove stray byebug

scmmishra committed:
feat: add i18n

scmmishra committed:
feat: retain query params in pagination

scmmishra committed:
Merge branch 'develop' into feature/10945-search-results-page

scmmishra commented:
@pranavrajs done PTAL

sojan-official committed:
Merge branch 'develop' into feature/10945-search-results-page

scmmishra committed:
Merge branch 'develop' into feature/10945-search-results-page

pranavrajs committed:
Merge branch 'develop' into feature/10945-search-results-page

pranavrajs commented:
@scmmishra I believe this issue might be related to the recent Turbo changes. Could you please take a look? The live chat widget on the Help Center is throwing an error during navigation.

<img width="561" alt="Screenshot 2025-06-11 at 7 10 06 PM" src="https://github.com/user-attachments/assets/20d1fba3-6589-48bc-afdf-d1fdbb1a2cbf" />

muhsin-k committed:
Merge branch 'develop' into feature/10945-search-results-page

sojan-official committed:
Merge branch 'develop' into feature/10945-search-results-page

scmmishra committed:
Merge branch 'develop' into feature/10945-search-results-page

scmmishra commented:
> I believe this issue might be related to the recent Turbo changes. Could you please take a look? The live chat widget on the Help Center is throwing an error during navigation.

@pranavrajs about this, there's a possibility you had an older build of the SDK, I couldn't reproduce it. Can you try `pnpm build:sdk` and see if it replicates?

sojan-official committed:
Merge branch 'develop' into feature/10945-search-results-page

muhsin-k committed:
Merge branch 'develop' into feature/10945-search-results-page

afanjul commented:
up

muhsin-k commented:
We wil

> Is any chance this simple PR could be released in any milestone? thanks

@afanjul We will test this feature and will get it merged this week.

muhsin-k committed:
Merge branch 'develop' into feature/10945-search-results-page

muhsin-k committed:
Update Gemfile.lock

muhsin-k committed:
Update Gemfile.lock

muhsin-k committed:
Merge branch 'develop' into feature/10945-search-results-page

afanjul commented:
Still not merged... @muhsin-k ?

stonecharioteer committed:
Merge branch 'develop' into feature/10945-search-results-page

sojan-official committed:
Merge origin/develop into feature/10945-search-results-page

afanjul committed:
Merge branch 'develop' into feature/10945-search-results-page

sojan-official committed:
Merge branch 'develop' into feature/10945-search-results-page

sojan-official committed:
Merge branch 'develop' into feature/10945-search-results-page

muhsin-k committed:
Merge branch 'develop' into feature/10945-search-results-page

muhsin-k committed:
Merge branch 'develop' into feature/10945-search-results-page

muhsin-k committed:
Merge branch 'develop' into feature/10945-search-results-page

sojan-official committed:
fix(portal): address search review feedback

sojan-official committed:
fix(portal): handle blank enterprise search queries

sojan-official committed:
fix(portal): preload search result categories

sojan-official committed:
Merge origin/develop into feature/10945-search-results-page

sojan-official committed:
fix(portal): apply search portal guards

sojan-official committed:
Merge branch 'develop' into feature/10945-search-results-page

sojan-official committed:
fix(portal): address search layout review notes

sojan-official committed:
refactor(portal): use native search form submission

sojan-official committed:
chore(portal): remove unused turbo rails gem

sojan-official committed:
Merge branch 'develop' into feature/10945-search-results-page

sojan-official committed:
refactor(portal): clean up search results view

sojan-official committed:
refactor(portal): align search view structure

sojan-official committed:
fix(portal): address search review comments

sojan-official committed:
Merge branch 'develop' into feature/10945-search-results-page

scmmishra commented:
@codex please review

sojan-official commented on enterprise/app/controllers/enterprise/public/api/v1/portals/search_controller.rb:
```diff
@@ -0,0 +1,9 @@
+  def search_articles
+    return super if @query.blank? || !@portal.account.feature_enabled?('help_center_embedding_search')
+
+    @articles = @articles.vector_search(search_params.merge(account_id: @portal.account_id, limit: nil))
```
this matches the current behavior in product so decided to leave it aside and take it up separately in a later pr.

scmmishra committed:
fix: prevent reserved slugs from shadowing help center routes

The /hc/:slug/:locale/search route would match an article whose slug
is "search", making it unreachable. Block "search" as an article slug
to keep article URLs resolving.

scmmishra commented on app/models/article.rb:
```diff
@@ -55,10 +55,14 @@ class Article < ApplicationRecord
   validates :author_id, presence: true
   validates :title, presence: true
   validates :content, presence: true, if: :published?
+  validates :slug, exclusion: { in: RESERVED_SLUGS }
```
Verified against the router: `:locale` in `hc/:slug/:locale/search` (routes.rb:592) is unconstrained and is registered before the article route, so `/hc/:slug/articles/search` resolves to `search#index`, not `articles#show` — the article is genuinely unreachable.

scmmishra committed:
Merge branch 'develop' into feature/10945-search-results-page

scmmishra committed:
feat: add more reserved slugs

scmmishra committed:
Merge branch 'develop' into feature/10945-search-results-page

scmmishra commented on app/views/public/api/v1/portals/search/index.html+documentation.erb:
```diff
@@ -0,0 +1,64 @@
+
+    <div class="grid grid-cols-1 gap-4">
+      <% @articles.each do |article| %>
+        <%= render 'public/api/v1/portals/documentation_layout/article_card',
+                   portal: @portal,
+                   article: article %>
```
The documentation view does not need this

scmmishra committed:
Merge branch 'develop' into feature/10945-search-results-page

afanjul commented:
its incredible and frustating that some people are trying to help in improving the chatwoot project and the dev team is totally abandoning the people, a easy PR to improve the UX of the platform **SENT on Mar 14, 2025!!!** and still not implemented, but not only that, I had to write like 4 to 5 comments trying to convince a team developer to f. merge the PR, but no way.  Seriously you are very annoying guys, you make that supporting open source project doesn't worth it at all.

scmmishra commented:
Hey @afanjul really sorry about this, and it's a great addition. We are a small team with customer commitments, internal roadmap and more to work through. That makes it difficult to split time on really good PRs from the community, we're experimenting with different SOPs within the team to make sure the good ones get enough attention. 

I know this can be frustrating, as it is dis-heartening for us that we are not able to provide enough attention to the PRs. I would thank you for your patience, and appreciate the fact that you're actively following up.

Rest assured, for this PR in particular, it's set to be merged this week. It's already marked for the upcoming release milestone. @sojan-official and I have pushed fixes and updates to bring it on parity with the recent changes. It's close to the finish line and will be merged soon
